diffusion/models/t2i_transformer.py

```diff
@@ -0,0 +1,459 @@
+        self.n_seq_params_per_block = self.model.num_features**2 * (4 + 2 * self.model.expansion_factor)
+        self.n_seq_params = self.n_seq_params_per_block * (self.model.num_layers - 1)
+        self.n_seq_params += 3 * (self.model.num_features**2)
+        self.n_last_layer_params = self.model.num_features**2 * (1 + 2 * self.model.expansion_factor)
+        # Params only on the input sequence
+        self.n_input_params = self.model.input_features * self.model.num_features
+        # Params only on the conditioning sequence
+        self.n_cond_params = self.model.conditioning_features * self.model.num_features
+
+        # Set up metrics
+        self.train_metrics = [MeanSquaredError()]
+        self.val_metrics = [MeanSquaredError()]
+
+        # Optional rng generator
+        self.rng_generator: Optional[torch.Generator] = None
+
+    def _apply(self, fn):
+        super(ComposerTextToImageMMDiT, self)._apply(fn)
+        self.latent_mean = fn(self.latent_mean)
+        self.latent_std = fn(self.latent_std)
+        return self
+
+    def set_rng_generator(self, rng_generator: torch.Generator):
+        """Sets the rng generator for the model."""
+        self.rng_generator = rng_generator
+
+    def flops_per_batch(self, batch) -> int:
+        batch_size = batch[self.image_key].shape[0]
+        height, width = batch[self.image_key].shape[2:]
+        input_seq_len = height * width / (self.patch_size**2 * self.downsample_factor**2)
+        cond_seq_len = batch[self.caption_key].shape[1]
+        seq_len = input_seq_len + cond_seq_len
+        # Calulate forward flops on full sequence excluding attention
+        param_flops = 2 * self.n_seq_params * batch_size * seq_len
+        # Last block contributes a bit less than other blocks
+        param_flops += 2 * self.n_last_layer_params * batch_size * input_seq_len
+        # Include input sequence params (comparatively small)
+        param_flops += 2 * self.n_input_params * batch_size * input_seq_len
+        # Include conditioning sequence params (comparatively small)
+        param_flops += 2 * self.n_cond_params * batch_size * cond_seq_len
+        # Include flops from adaln
+        param_flops += 2 * self.adaLN_params * batch_size
+        # Calculate flops for attention layers
+        attention_flops = 4 * self.model.num_layers * seq_len**2 * self.model.num_features * batch_size
+        return 3 * param_flops + 3 * attention_flops
+
+    def encode_image(self, image: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+        """Encode an image tensor with the autoencoder and patchify the latents."""
+        with torch.cuda.amp.autocast(enabled=False):
+            latents = self.autoencoder.encode(image.half())['latent_dist'].sample().data
+        # Scale and patchify the latents
+        latents = (latents - self.latent_mean) / self.latent_std
+        latent_patches, latent_coords = patchify(latents, self.patch_size)
```

Comment:
I think it makes sense to move the patching/unpatching logic into the `DiffusionTransformer` class. This way the Transformer logic is contained to the transformer and we wouldn't have to pass `coords` to the forward pass. Thoughts?

Reply:
I'd like to leave patching/unpatching out of `DiffusionTransformer` as it's specific to doing text to image training without precomputed latents, and in principle `DiffusionTransformer` is agnostic to the modality/preprocessing.